controller/initializer/initializer.go

```diff
@@ -0,0 +1,209 @@
+// Any error returned will cause the controller process to `Fatal`
+// The bool indicates whether the controller was enabled.
+type InitFunc func(manager.Manager) (enabled bool, err error)
+
+// InitOption configures how we set up initializer
+type InitOption interface {
+	apply(*options)
+}
+
+type options struct {
+	disableByDefault bool
+}
+type optionFunc func(*options)
+
+func (o optionFunc) apply(opt *options) {
+	o(opt)
+}
+
+// WithDisableByDefault disable controller by default
+func WithDisableByDefault() InitOption {
+	return optionFunc(func(o *options) {
+		o.disableByDefault = true
+	})
+}
+
+// Interface knows how to set up controllers with manager
+type Interface interface {
+	// Add add new controller setup function to initializer.
+	Add(controllerName string, setup InitFunc, options ...InitOption) error
+
+	// KnownControllers returns a slice of strings describing the ControllerInitialzier's known controllers.
+	KnownControllers() []string
+
+	// Enabled returns true if the controller is enabled.
+	Enabled(name string) bool
+
+	// SetupWithManager add all enabled controllers to manager
+	SetupWithManager(mgr manager.Manager) error
+
+	// BindFlag adds a flag for setting global feature gates to the specified FlagSet.
+	BindFlag(fs *pflag.FlagSet)
+}
+
+// New returns a new instance of initializer interface
+func New() Interface {
+	return &controllerInitializer{
+		initializers:     make(map[string]InitFunc),
+		all:              sets.NewString(),
+		enabled:          sets.NewString(),
+		disableByDefault: sets.NewString(),
+	}
+}
+
+var _ Interface = &controllerInitializer{}
+
+type controllerInitializer struct {
+	lock sync.Mutex
+
+	initializers     map[string]InitFunc
+	all              sets.String
+	enabled          sets.String
+	disableByDefault sets.String
+}
+
+func (m *controllerInitializer) Add(controllerName string, setup InitFunc, opts ...InitOption) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if m.all.Has(controllerName) {
+		return fmt.Errorf("controller %q has already been registered", controllerName)
+	}
+
+	opt := &options{}
+	for _, o := range opts {
+		o.apply(opt)
+	}
+
+	m.all.Insert(controllerName)
+
+	if opt.disableByDefault {
+		m.disableByDefault.Insert(controllerName)
+	} else {
+		m.enabled.Insert(controllerName)
+	}
+
+	m.initializers[controllerName] = setup
+	return nil
+}
+
+func (m *controllerInitializer) BindFlag(fs *pflag.FlagSet) {
+	all := m.all.List()
+	disabled := m.disableByDefault.List()
+	fs.Var(m, "controllers", fmt.Sprintf(""+
+		"A list of controllers to enable. '*' enables all on-by-default controllers, 'foo' enables the controller "+
+		"named 'foo', '-foo' disables the controller named 'foo'.\nAll controllers: %s\nDisabled-by-default controllers: %s",
+		strings.Join(all, ", "), strings.Join(disabled, ", ")))
+}
+
+// KnownControllers implements ControllerInitialzier.
+func (m *controllerInitializer) KnownControllers() []string {
```

Comment:
Method `Strinng` and `KnownControllers` is duplicated in some degree?
They can be sort in the same way before return.